File: frontend/src/api/online/protocol.ts
// オンライン対戦の WebSocket メッセージ（テトリス・ブロブチェイン共通）。
// backend/apps/tetris_online/protocol.py と同じ定義。

import type { RoomDto } from './types'

export type OnlineGame = 'tetris' | 'blob'

/** 相手画面に表示するための盤面。下の行から、各行 10 文字（'.' は空き、'G' はおじゃま、他はミノ） */
export type BoardRowsText = string[]

export interface PublicStats {
  pieces: number
  lines: number
  attack: number
}

export type ClientMessage =
  | { type: 'ready'; ready: boolean }
  | { type: 'state'; rows: BoardRowsText; stats: PublicStats; pending: number }
  | { type: 'attack'; lines: number }
  | { type: 'topout'; stats: PublicStats }
  | { type: 'ping' }

export type ServerMessage =
  | { type: 'welcome'; slot: number }
  | { type: 'room'; room: RoomDto }
  | { type: 'start'; seed: number; round: number; countdown_ms: number }
  | { type: 'opponent_state'; slot: number; rows: BoardRowsText; stats: PublicStats; pending: number }
  | { type: 'garbage'; from_slot: number; lines: number }
  | { type: 'end'; winner_slot: number; reason: 'topout' | 'disconnect'; room: RoomDto }
  | { type: 'error'; code: string; message: string }
  | { type: 'pong' }

export const CLOSE_ROOM_NOT_FOUND = 4004
export const CLOSE_ROOM_FULL = 4009

File: backend/apps/tetris_online/protocol.py
"""オンライン対戦の WebSocket メッセージ定義。

接続先: ws://<host>/ws/<tetris または blob>/rooms/<code>/?name=<ニックネーム>
すべて JSON で、"type" で種類を見分ける。フロント側の同じ定義は
frontend/src/api/online/protocol.ts。変えるときは両方を直すこと。

クライアント → サーバー
    {"type": "ready", "ready": true}                 準備完了（両者 ready で開始）
    {"type": "state", "rows": [...], "stats": {...}}  自分の盤面（相手画面の表示用。5〜10 回/秒程度）
    {"type": "attack", "lines": 4}                    相手に送る段数（相殺後）
    {"type": "topout", "stats": {...}}                自分がゲームオーバーになった
    {"type": "ping"}

サーバー → クライアント
    {"type": "welcome", "slot": 0}                    自分の番号（0 or 1）
    {"type": "room", "room": {...}}                   部屋の状態（参加者・ready・勝利数）
    {"type": "start", "seed": 123, "round": 1, "countdown_ms": 3000}
    {"type": "opponent_state", "slot": 1, "rows": [...], "stats": {...}}
    {"type": "garbage", "from_slot": 1, "lines": 4}   受け取るおじゃま
    {"type": "end", "winner_slot": 0, "reason": "topout", "room": {...}}
    {"type": "error", "code": "...", "message": "..."}
    {"type": "pong"}

※ 盤面の計算は各ブラウザが行い、サーバーは中継と勝敗の記録だけをする（不正対策はしていない）。
"""

COUNTDOWN_MS = 3000
MAX_ATTACK_PER_MESSAGE = 30
MAX_NAME_LENGTH = 20

# クライアント → サーバー
C_READY = "ready"
C_STATE = "state"
C_ATTACK = "attack"
C_TOPOUT = "topout"
C_PING = "ping"

# サーバー → クライアント
S_WELCOME = "welcome"
S_ROOM = "room"
S_START = "start"
S_OPPONENT_STATE = "opponent_state"
S_GARBAGE = "garbage"
S_END = "end"
S_ERROR = "error"
S_PONG = "pong"

# WebSocket を閉じるときのコード
CLOSE_ROOM_NOT_FOUND = 4004
CLOSE_ROOM_FULL = 4009
